### PR TITLE
Condition generator

### PR DESCRIPTION
**Pull request by davidleonfdez**
It's in draft state still. I've tried a first approach to condition the generator by prepending the corresponding labels (topic or form) to every poem. The idea is making the network learn that when a certain label appears at the beginning of a sequence, the generated text tends to follow certain patterns. 

I've chosen to put the labels at the beginning because in causal language models like gpt-2 every position depends only on the previous positions (and can only look at/attend to them), so this way every generated token depends on the labels. For the same reason, for classification with causal language models, the token used for classification is the last one, given that it is the only token that depends on the rest of the tokens.

I've experimented formatting the labels in different ways. You can see the results [here](https://github.com/Poems-AI/AI/blob/b5ee93d4db302d9c133931d0471873c7eb8a9fae/doc/results.md). Do you think it'd be better to create a different folder for results?
The metric to look at to assess the quality of the generated text, not taking into account the conditions, is "Best metadata-less val. loss".
To assess the conditional generation, you should look at the Topic/Form cond. loss/acc. A/B columns.

# Next steps

- Use an encoder-decoder architecture
  - This is the most obvious improvement and I'm 99% sure that it will work better than the current decoder-only architecture.
  - I am curious about how this will work as a zero-shot conditional generator, I mean: what happens if I feed to the encoder a condition it has never heard of? For instance, "This is a poem with sonnet form about love and expresses sad feelings".
    - Will it perform better for this matter if it's trained with a frozen pretrained encoder (i.e., the encoder weights are not updated), forcing the decoder to adapt to the encoder?
- Add a "conditional loss": send the output of the CLM to a classifier and compare the prediction with the expected label
  - This is quicker to implement now, I've added some code to trainer.py to compute and track multiple losses.
  - To keep thing simple, the classifier needs to be a GPT-2 too. Then, we can skip the logits.argmax and embedding indexing (not differentiable) and make a matrix multiplication between the CLM softmax(logits) (of size seq_len, vocab_size) and the clf input embedding weight matrix (of size vocab_size, hidden clf size). The result of the mm is a tensor where for each element of the sequence, we have a weighted sum of the embedding of each element of the vocab weighted by the predictions of the CLM; in conclusion, something like a soft indexing of the clf embedding depending on the tokens predicted by the CLM.
  - I'm not sure that this loss doesn't have any downsides because, while training, the output of the CLM is the prediction for every position given the previous tokens of the target, not a full sequence built token by token like it's done for inference. So we would be giving to the classifier something that maybe doesn't make sense. We can try anyway.
- Fill the missing labels with the prediction of classifiers
  - Not sure about the impact of this, but it's easy to implement now. It requires defining a different `label_func` to generate the inputs in notebooks/labeled_txt_inputs_generator.ipynb
- Other things that we could try while keeping the same architecture:
  - Train the model with two heads at the same time and a multi-objective loss function, one for the CLM and one for classification. My reasoning behind this, that might very well be wrong, is to incentivize the transformer to generate sequences easy to distinguish by category.
    - I don't think it's better than the encoder-decoder, so I'd skip this for now.

### CODE REVIEW

**Comment by davidleonfdez** [comment]
cc @ismaelfaro

**Comment by ismaelfaro** [comment]
I think that is good to have the result in the doc folder, the only thing is to be sure that if in the future we are going to have more results related with other models experiments, would be better identify if by the name of the file and add it explicitly like the title of the file, example os name from "results.md" >> "results_Condition_Generator.md".

**Comment by ismaelfaro** [comment]
I think that would be great add some extra info/comment in the main functions/classes

**Comment by davidleonfdez** [comment]
> I think that is good to have the result in the doc folder, the only thing is to be sure that if in the future we are going to have more results related with other models experiments, would be better identify if by the name of the file and add it explicitly like the title of the file, example os name from "results.md" >> "results_Condition_Generator.md".

Sure, that's a useful suggestion.

**Comment by davidleonfdez** [comment]
> I think that would be great add some extra info/comment in the main functions/classes

Yes, I have a tendency not to comment too much, apart from documentation in code, especially if I don't feel I can add much with a comment. In this case, there are some parameters, for example, that could be clearer if commented for sure. My approach is different for things like a public framework, and being this an open source project, yeah, I guess something in the middle like commenting the main functions sounds reasonable even for a little-comments-school folk like me.

Anyway, it's not my project, so I'll stick to the guidelines you'd like to set.

**Comment by davidleonfdez** [comment]
> * Fill the missing labels with the prediction of classifiers
>   
>   * Not sure about the impact of this, but it's easy to implement now. It requires defining a different `label_func` to generate the inputs in notebooks/labeled_txt_inputs_generator.ipynb

This is now done an tested, the improvement is small, if any.

**Comment by davidleonfdez** [comment]
> * Add a "conditional loss": send the output of the CLM to a classifier and compare the prediction with the expected label
>   
>   * This is quicker to implement now, I've added some code to trainer.py to compute and track multiple losses.
>   * To keep thing simple, the classifier needs to be a GPT-2 too. Then, we can skip the logits.argmax and embedding indexing (not differentiable) and make a matrix multiplication between the CLM softmax(logits) (of size seq_len, vocab_size) and the clf input embedding weight matrix (of size vocab_size, hidden clf size). The result of the mm is a tensor where for each element of the sequence, we have a weighted sum of the embedding of each element of the vocab weighted by the predictions of the CLM; in conclusion, something like a soft indexing of the clf embedding depending on the tokens predicted by the CLM.
>   * I'm not sure that this loss doesn't have any downsides because, while training, the output of the CLM is the prediction for every position given the previous tokens of the target, not a full sequence built token by token like it's done for inference. So we would be giving to the classifier something that maybe doesn't make sense. We can try anyway.

This is now implemented, but not evaluated nor integrated in the conditional generator training notebook (I haven't uploaded that notebook yet, I have it versioned in Kaggle for now, I will after I implement and add the encoder-decoder)

**Comment by ismaelfaro** [comment]
we can merge this pull request and open other for the new features to add, to keep all the PR small to be managed better.

**Comment by davidleonfdez** [comment]
Now that you mention it, this PR has gotten a bit out of scope hehe. I just marked it as ready.

**Comment by davidleonfdez** [comment]
Now that you mention it, this PR has gotten a bit out of scope hehe. I just marked it as ready.
